### ISSUE TITLE
Old TCP conn can panic OPTE

### ISSUE THREAD

**Issue by rzezeski** [oxidecomputer/opte#179]
I'm in the middle of shaving several yaks so this bug report will not be as crisp as I would like, but...

```
WARNING: mblk: 0xfffffe5897c99580
WARNING: ifid: TCP:10.0.0.10:56744:10.0.0.226:22
WARNING: flows: FlowTable { port_c: "opte2", name_c: "tcp_flows", limit: 8096, ttl: Ttl(60000), map: {InnerFlowId { proto: TCP, src_ip: Ip4(Ipv4Addr { inner: 172.30.0.6 }), src_port: 22, dst_ip: Ip4(Ipv4Addr { inner: 10.0.0.10 }), dst_port: 56744 }: FlowEntry { state: TcpFlowEntryState { inbound_ufid: Some(InnerFlowId { proto: TCP, src_ip: Ip4(Ipv4Addr { inner: 10.0.0.10 }), src_port: 56744, dst_ip: Ip4(Ipv4Addr { inner: 10.0.0.226 }), dst_port: 22 }), tcp_state: TcpFlowState { port_c: "opte2", tcp_state: Listen, guest_seq: None, guest_ack: None, remote_seq: Some(2422929007), remote_ack: None } }, hits: 0, last_hit: Moment { inner: 3795788574958 } }} }
WARNING: panicked at 'not yet implemented: bad packet: unexpected TCP segment dir: IN, flow: TCP:172.30.0.6:22:10.0.0.10:56744, state: LISTEN, flags: 0x18', /home/rpz/oxidecomputer/opte/opte/src/engine/port.rs:516:9

panic[cpu1]/thread=fffffe0079a12c20: 
panicked at 'not yet implemented: bad packet: unexpected TCP segment dir: IN, flow: TCP:172.30.0.6:22:10.0.0.10:56744, state: LISTEN, flags: 0x18', /home/rpz/oxidecomputer/opte/opte/src/engine/port.rs:516:9


fffffe0079a118a0 xde:rust_begin_unwind+8a ()
fffffe0079a118e0 xde:_ZN4core9panicking9panic_fmt17h7d7971299bdd0756E+2d ()
fffffe0079a119b0 xde:_ZN4opte6engine4port4Port14bad_packet_err17hbaf4f641a8504f55E+1cc ()
fffffe0079a11c30 xde:_ZN4opte6engine4port4Port7process17hecebe1beb9b8488aE+3487 ()
fffffe0079a12760 xde:xde_rx+4a4 ()
fffffe0079a127b0 mac:mac_promisc_dispatch_one+ac ()
fffffe0079a12810 mac:mac_promisc_client_dispatch+b4 ()
fffffe0079a12890 mac:mac_rx_srs_drain+1d2 ()
fffffe0079a12920 mac:mac_rx_srs_process+123 ()
fffffe0079a12970 mac:mac_rx_classify+7c ()
fffffe0079a129e0 mac:mac_rx_flow+58 ()
fffffe0079a12a40 mac:mac_rx_common+1de ()
fffffe0079a12aa0 mac:mac_rx+c6 ()
fffffe0079a12ae0 mac:mac_rx_ring+1f ()
fffffe0079a12b40 e1000g:e1000g_intr_work+264 ()
fffffe0079a12b70 e1000g:e1000g_intr_pciexpress+a3 ()
fffffe0079a12bc0 apix:apix_dispatch_by_vector+8c ()
fffffe0079a12c00 apix:apix_dispatch_lowlevel+1c ()
fffffe00799c19a0 unix:switch_sp_and_call+15 ()
fffffe00799c1a10 apix:apix_do_interrupt+ec ()
fffffe00799c1a20 unix:cmnint+c3 ()
fffffe00799c1ba0 unix:acpi_cpu_cstate+172 ()
fffffe00799c1bd0 unix:cpu_acpi_idle+94 ()
fffffe00799c1be0 unix:cpu_idle_adaptive+19 ()
fffffe00799c1c00 unix:idle+ae ()
fffffe00799c1c10 unix:thread_start+b ()
```

It seems OPTE does not deal well with a lingering half-open client connection to a guest. That is:

1. SSH into guest instance.
2. Stop/delete guest, but leave the SSH window on the client slide alone.
3. Recreate guest, it should get same external IP.
4. Go back to SSH session and try to input data.
5. The Omicron host should panic with the above.

The problem is that when creating a new TCP flow entry we don't check the flags on the first incoming packet, and basically assume its a SYN. We should be checking the flags, and in this particular case OPTE itself should probably respond with an RST on behalf of the guest since it has no knowledge of this connection anymore.

**Comment by rzezeski** [comment]
I ran the scenario described above and verified that the guest responds with a reset.

```
rpz@sector5 [~]
$ ssh -o StrictHostKeyChecking=no -o ControlMaster=no root@10.0.0.226
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@    WARNING: REMOTE HOST IDENTIFICATION HAS CHANGED!     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
IT IS POSSIBLE THAT SOMEONE IS DOING SOMETHING NASTY!
Someone could be eavesdropping on you right now (man-in-the-middle attack)!
It is also possible that a host key has just been changed.
The fingerprint for the ED25519 key sent by the remote host is
SHA256:Ossp0jNr4I0iAN8OeP8hBnZSM9vCIfGvQlL3PrGJqlo.
Please contact your system administrator.
Add correct host key in /Users/rpz/.ssh/known_hosts to get rid of this message.
Offending ECDSA key in /Users/rpz/.ssh/known_hosts:103
Password authentication is disabled to avoid man-in-the-middle attacks.
Keyboard-interactive authentication is disabled to avoid man-in-the-middle attacks.
UpdateHostkeys is disabled because the host key is not trusted.
Welcome to Alpine!

The Alpine Wiki contains a large amount of how-to guides and general
information about administrating Alpine systems.
See <http://wiki.alpinelinux.org/>.

You can setup the system with the command: setup-alpine

You may change this message by editing /etc/motd.

localhost:~# client_loop: send disconnect: Broken pipe
```

* If OPTE sees what looks like a data segment (ACK flag is set) for a non-existent flow it immediately moves the connection to established with the assumption that this is a valid connection that it has since forgotten about (say due to timeout expiration, or in this case an instance stop/start). At that point the guest will decide whether this is a legit connection or not. In this case the guest has been rebooted, so it has no idea about this connection anymore and replies with a reset, moving the connection immediately back to closed.
```
rpz@kalm:~/opte-dtrace$ pfexec ./opte-trace opte-tcp-flow-state.d 
PORT             CURR         NEW          FLOW
opte1            CLOSED       ESTABLISHED  TCP,172.30.0.5:22,10.0.0.10:50231
opte1            ESTABLISHED  CLOSED       TCP,172.30.0.5:22,10.0.0.10:50231
```

* Data segment from interacting with existing SSH session on client, followed by reset from guest.

```
rpz@kalm:~$ pfexec snoop -r -v -d net0 '(src 10.0.0.226) or (dst 10.0.0.226)'
Using device net0 (promiscuous mode)
                                                                            
IP:   ----- IP Header -----  
IP:                       
IP:   Version = 4                                                                        
IP:   Header length = 20 bytes         
IP:   Type of service = 0x48         
IP:         xxx. .... = 2 (precedence)
IP:         ...0 .... = normal delay
IP:         .... 1... = high throughput
IP:         .... .0.. = normal reliability
IP:         .... ..0. = not ECN capable transport
IP:         .... ...0 = no ECN congestion experienced
IP:   Total length = 88 bytes    
IP:   Identification = 0                   
IP:   Flags = 0x4                        
IP:         .1.. .... = do not fragment                                                  
IP:         ..0. .... = last fragment                                                    
IP:   Fragment offset = 0 bytes                                                          
IP:   Time to live = 64 seconds/hops                                                     
IP:   Protocol = 6 (TCP)                                                                 
IP:   Header checksum = 256d 
IP:   Source address = 10.0.0.10, 10.0.0.10
IP:   Destination address = 10.0.0.226, 10.0.0.226      
IP:   No options                          
IP:                                      
TCP:  ----- TCP Header -----              
TCP:                           
TCP:  Source port = 50231        
TCP:  Destination port = 22                                                              
TCP:  Sequence number = 1645922004                                                       
TCP:  Acknowledgement number = 2926029014
TCP:  Data offset = 32 bytes
TCP:  Flags = 0x18      
TCP:        0... .... = No ECN congestion window reduced
TCP:        .0.. .... = No ECN echo
TCP:        ..0. .... = No urgent pointer
TCP:        ...1 .... = Acknowledgement
TCP:        .... 1... = Push
TCP:        .... .0.. = No reset
TCP:        .... ..0. = No Syn
TCP:        .... ...0 = No Fin
TCP:  Window = 2048
TCP:  Checksum = 0xe1be
TCP:  Urgent pointer = 0
TCP:  Options: (12 bytes)
TCP:    - No operation
TCP:    - No operation
TCP:    - TS Val = 365213668, TS Echo = 2310761331
TCP:  

IP:   ----- IP Header -----
IP:   
IP:   Version = 4
IP:   Header length = 20 bytes
IP:   Type of service = 0x48
IP:         xxx. .... = 2 (precedence)
IP:         ...0 .... = normal delay
IP:         .... 1... = high throughput
IP:         .... .0.. = normal reliability
IP:         .... ..0. = not ECN capable transport
IP:         .... ...0 = no ECN congestion experienced
IP:   Total length = 40 bytes
IP:   Identification = 0
IP:   Flags = 0x4
IP:         .1.. .... = do not fragment
IP:         ..0. .... = last fragment
IP:   Fragment offset = 0 bytes
IP:   Time to live = 64 seconds/hops
IP:   Protocol = 6 (TCP)
IP:   Header checksum = 259d
IP:   Source address = 10.0.0.226, 10.0.0.226
IP:   Destination address = 10.0.0.10, 10.0.0.10
IP:   No options
IP:   
TCP:  ----- TCP Header -----
TCP:  
TCP:  Source port = 22
TCP:  Destination port = 50231 
TCP:  Sequence number = 2926029014
TCP:  Acknowledgement number = 0
TCP:  Data offset = 20 bytes
TCP:  Flags = 0x04
TCP:        0... .... = No ECN congestion window reduced
TCP:        .0.. .... = No ECN echo
TCP:        ..0. .... = No urgent pointer
TCP:        ...0 .... = No acknowledgement
TCP:        .... 0... = No push
TCP:        .... .1.. = Reset
TCP:        .... ..0. = No Syn
TCP:        .... ...0 = No Fin
TCP:  Window = 0
TCP:  Checksum = 0x7f69
TCP:  Urgent pointer = 0
TCP:  No options
TCP:
```

**Comment by rzezeski** [comment]
Addressed in fb01edd13197fbc8626f852397ce2090a8502cbe.
